util/tls/tls_socket.cc

```diff
@@ -505,13 +505,173 @@ void TlsSocket::OnRecv(const RecvNotification& rn, const OnRecvCb& recv_cb) {
 }
 
 io::Result<size_t> TlsSocket::TrySend(io::Bytes buf) {
-  LOG(DFATAL) << "Not implemented";
-  return 0;
+  iovec vec[1];
+  vec[0].iov_base = const_cast<uint8_t*>(buf.data());
+  vec[0].iov_len = buf.size();
+  return TrySend(vec, 1);
+}
+
+void TlsSocket::AdvanceIovec(iovec*& iov, uint32_t& len, size_t bytes_to_advance) {
+  DCHECK_NE(iov, nullptr);
+  DCHECK_GT(len, 0u);
+
+  while ((len > 0) && (bytes_to_advance > 0)) {
+    if (bytes_to_advance >= iov->iov_len) {  // case 1: remove the entry
+      bytes_to_advance -= iov->iov_len;
+      ++iov;
+      --len;
+    } else {  // case 2: adjust the current entry's start and length
+      iov->iov_base = static_cast<char*>(iov->iov_base) + bytes_to_advance;
+      iov->iov_len -= bytes_to_advance;
+      bytes_to_advance = 0;
+      break;
+    }
+  }
+  DCHECK_EQ(bytes_to_advance, 0u) << "AdvanceIovec logic error: unconsumed bytes remaining";
 }
 
 io::Result<size_t> TlsSocket::TrySend(const iovec* v, uint32_t len) {
-  LOG(DFATAL) << "Not implemented";
-  return 0;
+  if (len == 0)
+    return 0;  // nothing to send
+  bool has_data{false};
+  for (size_t i{}; i < len; ++i) {
+    if (v[i].iov_len > 0) {
+      has_data = true;
+      break;
+    }
+  }
+  if (!has_data)
+    return 0;  // nothing to send
```

Comment:
can be factored out into a helper function, and lets DCHECK as well, as I do not see why should we allow callers to call this function with no data